src/s3fs.cpp

```diff
@@ -2746,6 +2781,55 @@ static int readdir_multi_head(const char* path, const S3ObjList& head, void* buf
         }
     }
 
+    // [NOTE]
+    // Objects that could not be found by HEAD request may exist only
+    // as a path, so search for objects under that path.(a case of no dir object)
+    //
+    if(support_compat_dir && !notfound_param.notfound_list.empty()){      // [NOTE] not need to lock to access this here.
+        // dummy header
+        mode_t dirmask = umask(0);      // macos does not have getumask()
+        umask(dirmask);
+
+        headers_t   dummy_header;
+        dummy_header["Content-Type"]     = std::string("application/x-directory");          // directory
+        dummy_header["x-amz-meta-uid"]   = str(is_s3fs_uid ? s3fs_uid : geteuid());
+        dummy_header["x-amz-meta-gid"]   = str(is_s3fs_gid ? s3fs_gid : getegid());
+        dummy_header["x-amz-meta-mode"]  = str(S_IFDIR | (~dirmask & (S_IRWXU | S_IRWXG | S_IRWXO)));
+        dummy_header["x-amz-meta-atime"] = str(0);
+        dummy_header["x-amz-meta-ctime"] = str(0);
+        dummy_header["x-amz-meta-mtime"] = str(0);
+
+        for(s3obj_list_t::iterator reiter = notfound_param.notfound_list.begin(); reiter != notfound_param.notfound_list.end(); ++reiter){
+            int dir_result;
+            if(0 == (dir_result = directory_empty(reiter->c_str()))){
```

Comment:
Does this call `directory_empty` and thus `list_bucket` once per not found object?  This will have bad performance.

Reply:
Yes, as you pointed out, this will degrade performance.
(However, only when the `compat_dir` option is specified and the directory is not found.)

For simple way, conditions that affect performance can be mitigated by specifying a level in the option of `compat_dir`(ex, old_type_dir, no_dir_type, both...) or preparing another option.

In the future, performance can be improved by not calling `directory_empty` and executing Head requests in parallel or run `directory_empty` in parallel.

However, this performance loss may not have a widespread impact, given that `compat_dir` is disabled by `default`.
(Since it is accumulated in the stat cache, this serial requests does not occur every time.)